relations/src/shielder/mod.rs

```diff
@@ -55,53 +56,33 @@ fn convert_account(front: [u8; 32]) -> CircuitField {
 }
 
 fn check_merkle_proof(
-    merkle_root: Result<&BackendMerkleRoot, SynthesisError>,
-    leaf_index: Result<&BackendLeafIndex, SynthesisError>,
-    leaf_bytes: Vec<UInt8<CircuitField>>,
+    merkle_root: FpVar,
+    path_shape: PathShapeVar,
+    leaf: FpVar,
     path: BackendMerklePath,
     max_path_len: u8,
     cs: ConstraintSystemRef<CircuitField>,
 ) -> Result<(), SynthesisError> {
     if path.len() > max_path_len as usize {
         return Err(UnconstrainedVariable);
     }
+    if path_shape.len() != max_path_len as usize {
+        return Err(UnconstrainedVariable);
+    }
 
-    let zero = CircuitField::zero();
-
-    let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || merkle_root)?;
-    let mut leaf_index = FpVar::new_witness(ns!(cs, "leaf index"), || leaf_index)?;
-
-    let mut current_hash_bytes = leaf_bytes;
-    let mut hash_bytes = vec![current_hash_bytes.clone()];
+    let mut current_note = leaf;
+    let zero_note = CircuitField::zero();
 
     for i in 0..max_path_len {
         let sibling = FpVar::new_witness(ns!(cs, "merkle path node"), || {
-            Ok(path.get(i as usize).unwrap_or(&zero))
+            Ok(path.get(i as usize).unwrap_or(&zero_note))
```

Comment:
Since we use `i as usize` everywhere, you could cast in the `for` loop above 
```rust
for i in 0..max_path_len as usize
```

Reply:
done